framework/db/QueryBuilder.php

```diff
@@ -145,6 +166,43 @@ public function build($query, $params = [])
         return [$sql, $params];
     }
 
+    /**
+     * Builds given $expression
+     *
+     * @param ExpressionInterface $expression the expression to be built
+     * @param array $params the parameters to be bound to the generated SQL statement. These parameters will
+     * be included in the result with the additional parameters generated during the expression building process.
+     * @return string the SQL statement that will not be neither quoted nor encoded before passing to DBMS
+     * @see ExpressionInterface
+     * @see ExpressionBuilderInterface
+     * @see expressionBuilders
+     * @since 2.0.14
+     * @throws InvalidParamException when $expression building is not supported by this QueryBuilder.
+     */
+    public function buildExpression(ExpressionInterface $expression, &$params = [])
+    {
+        $className = get_class($expression);
+        if (!isset($this->expressionBuilders[$className])) {
+            foreach ($this->expressionBuilders as $expressionClass => $builderClass) {
+                if (is_subclass_of($expression, $expressionClass)) {
+                    $this->expressionBuilders[$className] = $builderClass;
```

Comment:
it's a bit counter intuitive that behavior depends on the order of execution
imagine:
we have `$this->expressionBuilders = ['…\ParentExpression' => '…\ConcreteBuilder']
we have `class ChildExpression extends …\ParentExpression`

if we call `->buildExpression($childExpression)` and then `->buildExpression($parentExpression)` then we create separate builder instances for each class

if we call `->buildExpression($parentExpression)` and then `->buildExpression($childExpression)` then we create one instance that is shared between classes

Reply:
if we have
`$this->expressionBuilders = ['…\GrandParentExpression' => '…\Builder1', '…\ParentExpression' => '…\Builder2']`
and we pass instance of `…\ChildExpression` to be built, then we would use Builder1, but it seems Builder 2 would be more suitable

it may be problem as setExpressionBuilders is only appending to the list, it's worth documenting at least

Reply:
Yes, that my be confusing. Of course we can go deeper and find the most suitable builder with `class_parents` and `class_impements`, but it will affect performance and increase code complexity. Do we need it?

I'm OK with documenting this behavior.

Reply:
Considering `array_merge` appends builders to the end of list I think builders should be processed in reverse order which allows to solve the problem.